RecoEgamma/PhotonIdentification/src/PhotonMVABasedHaloTagger.cc

```diff
@@ -0,0 +1,424 @@
+
+    if (thresholds == nullptr) {
+      throw cms::Exception("EmptyPFRechHitThresCollection")
+          << "In PhotonMVABasedHaloTagger::calphoClusCoordinECAL, EcalPFRecHitThresholds cannot be = nulptr";
+    }
+
+    float rhThres = 0.0;
+    rhThres = (*thresholds)[det];
+
+    if (rhE <= rhThres)
+      continue;
+
+    if (phoSCEta * rhEta < 0)
+      continue;
+
+    double dR2 = reco::deltaR2(rhEta, rhPhi, phoSCEta, phoSCPhi);
+
+    if (dR2 < 0.2 * 0.2) {
+      ecalClusX_ += rhX * rhE;
+      ecalClusY_ += rhY * rhE;
+      ecalClusZ_ += rhZ * rhE;
+      ecalClusE_ += rhE;
+      ecalClusNhits_++;
+    }
+  }  //for(int ih=0; ih<nMatchedEErh[ipho]; ih++)
+
+  if (ecalClusNhits_ > 0) {  //should always be > 0 for an EM cluster
+    ecalClusX_ = ecalClusX_ / ecalClusE_;
+    ecalClusY_ = ecalClusY_ / ecalClusE_;
+    ecalClusZ_ = ecalClusZ_ / ecalClusE_;
+  }  //if(ecalClusNhits_>0)
+}
+
+void PhotonMVABasedHaloTagger::calmatchedHBHECoordForBothHypothesis(const CaloGeometry* geo,
+                                                                    const reco::Photon* pho,
+                                                                    const HBHERecHitCollection& HBHERecHits) {
+  hcalClusX_samedPhi_ = 0;
+  hcalClusY_samedPhi_ = 0;
+  hcalClusZ_samedPhi_ = 0;
+
+  hcalClusNhits_samedPhi_ = 0;
+  hcalClusE_samedPhi_ = 0;
+
+  hcalClusX_samedR_ = 0;
+  hcalClusY_samedR_ = 0;
+  hcalClusZ_samedR_ = 0;
+  hcalClusNhits_samedR_ = 0;
+  hcalClusE_samedR_ = 0;
+
+  double phoSCEta = pho->superCluster()->eta();
+  double phoSCPhi = pho->superCluster()->phi();
+
+  // Loop over HBHERecHit's
+  for (const auto& hbherechit : HBHERecHits) {
+    HcalDetId det = hbherechit.id();
+    const GlobalPoint& rechitPoint = geo->getPosition(det);
+
+    double rhEta = rechitPoint.eta();
+    double rhPhi = rechitPoint.phi();
+    double rhX = rechitPoint.x();
+    double rhY = rechitPoint.y();
+    double rhZ = rechitPoint.z();
+    double rhE = hbherechit.energy();
+
+    int depth = det.depth();
+
+    if ((det.subdet() == HcalBarrel and (depth < 1 or depth > int(recHitEThresholdHB_.size()))) or
+        (det.subdet() == HcalEndcap and (depth < 1 or depth > int(recHitEThresholdHE_.size()))))
+      edm::LogWarning("PhotonMVABasedHaloTagger")
+          << " hit in subdet " << det.subdet() << " has an unaccounted for depth of " << depth << "!!";
+
+    const bool goodHBe = det.subdet() == HcalBarrel and rhE > recHitEThresholdHB_[depth - 1];
+    const bool goodHEe = det.subdet() == HcalEndcap and rhE > recHitEThresholdHE_[depth - 1];
+    if (!(goodHBe or goodHEe))
+      continue;
+
+    if (phoSCEta * rhEta < 0)
+      continue;  ///Should be on the same side of Z
+
+    double rho2 = pow(rhX, 2) + pow(rhY, 2);
+    double dPhi = deltaPhi(phoSCPhi, rhPhi);
+
+    bool isRHBehindECAL = false;
+
+    double dRho2 = pow(rhX - ecalClusX_, 2) + pow(rhY - ecalClusY_, 2);
+
+    if (rho2 >= 31 * 31 && rho2 <= 172 * 172 && dRho2 <= 26 * 26 &&
```

Comment:
usually we ask such constant to be defined in a header as constexpr, even if used just once.